crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,248 @@
+
+    #[derive(Debug)]
+    pub struct FileReader {
+        inner: web_sys::FileReader,
+    }
+
+    impl FileReader {
+        pub fn new() -> FileReader {
+            FileReader {
+                inner: web_sys::FileReader::new().unwrap_throw(),
+            }
+        }
+
+        pub fn read_as_string(self, blob: &impl BlobLike) -> ReadAsString {
+            let (closure, rx) = self.read_as(
+                blob,
+                |js| js.as_string().unwrap_throw(),
+                web_sys::FileReader::read_as_text,
+            );
+            ReadAsString {
+                inner: self.inner,
+                _closure: closure,
+                receiver: rx,
+            }
+        }
+
+        pub fn read_as_data_url(self, blob: &impl BlobLike) -> ReadAsDataUrl {
+            let (closure, rx) = self.read_as(
+                blob,
+                |js| js.as_string().unwrap_throw(),
+                web_sys::FileReader::read_as_data_url,
+            );
+
+            ReadAsDataUrl {
+                inner: self.inner,
+                _closure: closure,
+                receiver: rx,
+            }
+        }
+
+        pub fn read_as_array_buffer(self, blob: &impl BlobLike) -> ReadAsArrayBuffer {
+            let (closure, rx) = self.read_as(
+                blob,
+                std::convert::Into::into,
+                web_sys::FileReader::read_as_array_buffer,
+            );
+
+            ReadAsArrayBuffer {
+                inner: self.inner,
+                _closure: closure,
+                receiver: rx,
+            }
+        }
+
+        fn read_as<T, F, G>(
+            &self,
+            blob: &impl BlobLike,
+            convert: F,
+            start_read: G,
+        ) -> (
+            wasm_bindgen::closure::Closure<FnMut() + 'static>,
+            futures::sync::oneshot::Receiver<T>,
+        )
+        where
+            T: 'static,
+            F: Fn(wasm_bindgen::JsValue) -> T + 'static,
+            G: Fn(&web_sys::FileReader, &web_sys::Blob) -> Result<(), wasm_bindgen::JsValue>,
+        {
+            let (tx, rx) = futures::sync::oneshot::channel();
+            let reader = self.inner.clone();
+            let closure = Closure::once(move || {
+                let _ = reader.result().map(|js_value| {
+                    let _ = tx.send(convert(js_value));
+                });
+            });
+            let function = closure.as_ref().dyn_ref().unwrap_throw();
+            self.inner.clone().set_onload(Some(&function));
+            start_read(&self.inner, blob.as_raw()).unwrap_throw();
+            (closure, rx)
+        }
+    }
+
+    macro_rules! future_impls {
+       ($($name:ident : $item:ty),*) => ($(
+            pub struct $name {
+                receiver: oneshot::Receiver<$item>,
+                _closure: Closure<FnMut()>,
+                inner: web_sys::FileReader,
+            }
+
+            impl Future for $name {
```

Comment:
I think we should seriously consider using Futures 0.3 instead.

The `Future` trait is already stabilized, and Futures 0.3 will be available on stable soon.

I'm also okay with delaying the `futures` submodule until after Futures 0.3 is available on stable (only releasing the `callbacks` submodule).